src/basic/glob-util.c

```diff
@@ -71,3 +71,22 @@ int glob_extend(char ***strv, const char *path, int flags) {
 
         return strv_extend_strv(strv, g.gl_pathv, false);
 }
+
+int glob_non_glob_prefix(const char *path, char **ret) {
+        /* Return the path of the path that has no glob characters. */
+
+        size_t n = strcspn(path, GLOB_CHARS);
+
+        if (path[n] != '\0')
+                while ( n > 0 && path[n-1] != '/')
+                        n--;
+
+        if (n == 0)
+                return -ENOENT;
+
+        char *ans = strndup(path, n);
+        if (!ans)
+                return -ENOMEM;
+        *ret = ans;
+        return 0;
```

Comment:
maybe return this only if the result is still a valid path?

another idea would be to implement this as a loop around path_extract_directory() that always checks whether the current result still contains globs, and leaves once there aren't any

Reply:
> maybe return this only if the result is still a valid path?

I think this is what happens already. We strip everything until the last slash before the first glob character. Empty strings are rejected, so whatever remains is a valid path. (We should not filter relative or non-absolute paths here, and I don't think there's anything else we *could* filter.)

I implemented the other suggestions, so I'll upgrade the flag after pushing.

Reply:
> as a loop around path_extract_directory()

This would work too, but I think it would be better… For expressions with globs not at the end we would do path_extract_directory() a few times, and check for globs a few times too. With the current implementation we jump straight to the chase. It doesn't matter *too* much if we spend some extra cycles on this, but I don't think there's a reason to change.